dsc/workflows/base/__init__.py

```diff
@@ -371,119 +327,108 @@ def process_result_messages(self) -> None:
+                result_message_map[item_identifier] = message
+            except InvalidSQSMessageError:
+                logger.exception(f"Failure parsing message '{message["MessageId"]}'")
                 continue
 
-            item_identifier = message_attributes["PackageID"]["StringValue"]
+        sqs_results_summary["received_messages"] = len(result_message_map)
 
-            # get record from ItemSubmissionDB
-            try:
-                item_submission_record = ItemSubmissionDB.get(
-                    hash_key=self.batch_id, range_key=item_identifier
-                )
-            except DoesNotExist:
-                logger.warning(
-                    "Record "
-                    f"{ITEM_SUBMISSION_LOG_STR.format(batch_id=self.batch_id,
-                                      item_identifier=item_identifier)}"
-                    "not found. Verify that it been reconciled."
-                )
+        # retrieve item submissions from batch
+        for item_submission in ItemSubmission.get_batch(self.batch_id):
+            log_str = ITEM_SUBMISSION_LOG_STR.format(
+                batch_id=self.batch_id, item_identifier=item_submission.item_identifier
+            )
+            if item_submission.status == ItemSubmissionStatus.INGEST_SUCCESS:
+                logger.debug(f"Record {log_str} already ingested, skipping")
                 continue
 
-            current_status = item_submission_record.status
+            item_submission.ingest_attempts += 1
 
-            logger.info(
-                "Received result message for record "
-                f"{ITEM_SUBMISSION_LOG_STR.format(batch_id=item_submission_record.batch_id,
-                                      item_identifier=item_submission_record.item_identifier)}"
-            )
+            result_message = result_message_map.get(item_submission.item_identifier)
 
-            try:
-                parsed_message_body = self._parse_result_message_body(message_body)
-            except InvalidSQSMessageError as exception:
-                logger.error(  # noqa: TRY400
-                    f"Failed to parse 'Body' from {message_id}:{exception}"
-                )
+            # set status to ingest_unknown if no message exists with item_identifier
+            if not result_message:
+                item_submission.status = ItemSubmissionStatus.INGEST_UNKNOWN
                 sqs_results_summary["ingest_unknown"] += 1
+                item_submission.upsert_db()
+                logger.info(f"Unable to determine ingest status for record {log_str}")
+                continue
 
-                new_status = ItemSubmissionStatus.INGEST_UNKNOWN
-
-                item_submission_record.update(
-                    actions=[ItemSubmissionDB.status_details.set(str(exception))]
+            # update item submission status based on ingest result
+            result_message_body = json.loads(result_message["Body"])
```

Comment:
Could we do this JSON loading during the result message parsing, e.g. as part of the proposed `retrieve_and_parse_result_messages()` sub-method?

Reply:
Update: see above about possible `ResultMessage` `@dataclass`.